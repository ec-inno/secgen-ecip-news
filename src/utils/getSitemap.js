import { unflatten } from 'flat';

/**
 * Provides a nested structure representing a sitemap, based on main menu and pages.
 * @param {Object} data Pages' graphql queries result.
 * @param {String} locale
 */
const getSitemap = ({ data, locale }) => {
  const listPrepare = [];
  let listGroupPrepare = {};

  // Menu information about `fetched_alias`, aliased as `href` below, contains site folder we don't need.
  const drupalSiteFolder = process.env.GATSBY_DRUPAL_SITE_FOLDER
    ? `/${process.env.GATSBY_DRUPAL_SITE_FOLDER}/`
    : '';

  const menu = data.allMenu.edges.length
    ? data.allMenu.edges
        .map(({ node }) => node)
        .map(link => ({
          title: link.title,
          hrefFormatted: link.href
            .replace(drupalSiteFolder, '')
            .replace(`${locale}/`, `/${locale}/`),
          external: link.external,
        }))
        .map(link => ({
          ...link,
          href:
            link.hrefFormatted === locale ? `/${locale}` : link.hrefFormatted,
        }))
    : [];

  const pages =
    data.allNodeOePage && data.allNodeOePage.edges
      ? data.allNodeOePage.edges
          .map(({ node }) => ({
            title: node.title,
            href: `/${locale}${node.path.alias}`,
          }))
          // Keep only items which are not already present in the menu.
          .filter(link => !menu.find(m => m.href === link.href))
      : [];

  const others =
    data.allSitePage && data.allSitePage.edges
      ? data.allSitePage.edges.map(({ node }) => ({
          title:
            node.context && node.context.title ? node.context.title : 'test',
          href: node.path,
        }))
      : [];

  listPrepare.push(...menu);
  listPrepare.push(...pages);
  listPrepare.push(...others);

  const home = listPrepare.shift();

  // Prepare list for a nested structure.
  listPrepare
    .map(item => {
      const path = item.href.split('/').filter(a => a);
      path.shift();
      const href = path.join('.');

      return {
        hrefNew: href,
        ...item,
      };
    })
    .filter(item => item.hrefNew)
    .forEach(item => {
      if (!listGroupPrepare[item.hrefNew]) {
        listGroupPrepare[item.hrefNew] = item;
      }
    });

  // Add back the link to Home.
  /* eslint-disable-next-line */
  listGroupPrepare = Object.assign({ [home.title]: home }, listGroupPrepare);

  // Keep a set of link attributes, the rest will be treated as children for rendering nested unordered lists.
  const listFormatted = {};

  Object.keys(listGroupPrepare).forEach(key => {
    const { title, href, external: ex } = listGroupPrepare[key];
    listFormatted[key] = { title, href, external: ex || false };
  });

  const list = unflatten(listFormatted);

  return list;
};

export default getSitemap;
