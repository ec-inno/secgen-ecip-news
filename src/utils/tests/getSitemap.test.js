import getSitemap from '../getSitemap';

describe('getSitemap utlity', () => {
  it('returns an empty object on empty input', () => {
    const result = getSitemap();
    expect(result).toEqual({});
  });

  it('returns a structure of links', () => {
    // Mock is a result of the executed query in src/pages/sitemap
    const data = require('./stubs/sitemap/data');
    const locale = 'en';
    const result = getSitemap({ data, locale });
    expect(result).toMatchSnapshot();
  });
});
