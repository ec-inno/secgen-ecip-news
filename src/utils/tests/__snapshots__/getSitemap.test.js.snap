// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSitemap utlity returns a structure of links 1`] = `
Object {
  "Home": Object {
    "external": false,
    "href": "/multisite/ecip/en",
    "title": "Home",
  },
  "contact": Object {
    "external": false,
    "href": "/en/contact",
    "title": "Contact",
  },
  "faq": Object {
    "external": false,
    "href": "/en/faq",
    "title": "FAQ",
  },
  "find-initiative": Object {
    "external": false,
    "href": "/en/find-initiative",
    "title": "Find initiative",
  },
  "how-it-works": Object {
    "data-protection": Object {
      "external": false,
      "href": "/en/how-it-works/data-protection",
      "title": "Data protection",
    },
    "external": false,
    "href": "/en/how-it-works",
    "implementation-national-level": Object {
      "external": false,
      "href": "/en/how-it-works/implementation-national-level",
      "title": "Implementation at national level",
    },
    "regulatory-framework": Object {
      "external": false,
      "href": "/en/how-it-works/regulatory-framework",
      "title": "Regulatory framework for the European Citizens' Initiative",
    },
    "title": "How it works",
  },
  "how-to-start": Object {
    "competences": Object {
      "external": false,
      "href": "/en/how-to-start/competences",
      "title": "Can your idea be a citizens' initiative?",
    },
    "external": false,
    "href": "/en/how-to-start",
    "prepare-system": Object {
      "external": false,
      "href": "/en/how-to-start/prepare-system",
      "title": "Prepare your online collection system",
    },
    "title": "How to start an initiative",
  },
  "news": Object {
    "external": false,
    "href": "/en/news",
    "title": "News",
  },
}
`;
